pkg/leech/pipeline.go

```diff
@@ -0,0 +1,274 @@
+}
+
+// ingestLogsFn is an object that implements beams "DoFn" interface to
+// provide the main processing of the event.
+type ingestLogsFn struct {
+	pipe   *Pipeline
+	client *http.Client
+}
+
+// BeamInit handles preregistration of pipeline shapes with their associated
+// type signature and shape to speed up runtime execution.
+func BeamInit() {
+	register.DoFn1x1[EventRecord, LeechRecord](&ingestLogsFn{})
+}
+
+// New creates a new instance of the leech Pipeline.
+func New(ctx context.Context, cfg *Config) (*Pipeline, error) {
+	store, err := NewObjectStore(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create object store client: %w", err)
+	}
+	pk, err := readPrivateKey(cfg.GitHubPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	ghAppConfig := githubapp.NewConfig(cfg.GitHubAppID, cfg.GitHubInstallID, pk)
+	ghApp := githubapp.New(ghAppConfig)
+
+	return &Pipeline{
+		cfg:     cfg,
+		ghApp:   ghApp,
+		storage: store,
+	}, nil
+}
+
+// Prepare connects the internal pipeline implementation to the beam
+// scope object associated with the primary beam Pipeline.
+func (pipe *Pipeline) Prepare(scope beam.Scope) {
+	// BigQuery table notation is not consistent so it needs represented in a few different
+	// formats to appease the BigQuery client.
+	eventsTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.EventsProjectID, pipe.cfg.EventsTable)
+	leechTableDotNotation := fmt.Sprintf("%s.%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	leechTableColonNotation := fmt.Sprintf("%s:%s", pipe.cfg.LeechProjectID, pipe.cfg.LeechTable)
+	batchSize := pipe.cfg.BatchSize
+	client := http.Client{Timeout: 5 * time.Minute}
+
+	var event EventRecord
+	query := fmt.Sprintf(sourceQuery, eventsTableDotNotation, leechTableDotNotation, batchSize, batchSize)
+	// step 1: query BigQuery for unprocessed events
+	col := bigqueryio.Query(scope, pipe.cfg.LeechProjectID, query, reflect.TypeOf(event), bigqueryio.UseStandardSQL())
+	// step 2: process the events in parallel, ingesting logs
+	res := beam.ParDo(scope, &ingestLogsFn{pipe: pipe, client: &client}, col)
+	// step 3: write all of the results back to BigQuery
+	bigqueryio.Write(scope, pipe.cfg.LeechProjectID, leechTableColonNotation, res)
+}
+
+// readPrivateKey reads a PEM encoded private key from a string.
+func readPrivateKey(privateKeyContent string) (*rsa.PrivateKey, error) {
+	parsedKey, _, err := jwk.DecodePEM([]byte(privateKeyContent))
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode PEM formated key:  %w", err)
+	}
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert to *rsa.PrivateKey (got %T)", parsedKey)
+	}
+	return privateKey, nil
+}
+
+// ProcessElement is a DoFn implementation that reads workflow logs from GitHub
+// and stores them in Cloud Storage.
+func (f *ingestLogsFn) ProcessElement(event EventRecord) LeechRecord {
+	ctx := context.Background()
```

Comment:
Not familiar with Dataflow but this feels weird. Any better way to handle the context?

Reply:
There is no easy way to pass the context around in this case and you really wouldn't want to. 

Don't think of this as a typical Go program that runs self contained. This function is basically a UDF that will be serialized to N workers and the `runner` will schedule messages to be processed by that worker via this function. That execution is in a new context from the initial one created on the `runner`. 

This might look like a single program that executes but what's really happening is a pipeline definition is being compiled and then handed off to a `runner` to execute. The runner has no knowledge of how the pipeline was created it just knows the computational units it needs to execute based on the pipeline definition.

The call to `beam.Run()` in `main.go` is telling the runner - here is a pipeline definition I just built for you, please go execute it. When using the local runner this looks like a few threads and a typical Go application but in Dataflow or any other distributed runner it is pushed to multiple workers. The first stage is executed to build a collection of EventRecords to process. The runner then determines how many workers it wants to use to process those and splits the collection into parts, serializes those to worker machines along with the function definition which is then executed. Results are collected back on the runner and stored in another collection which is then sent to the final step which is a BigQuery batch write operation. 

Each of those operations is a self contained unit of work that is executed independently.

We can chat more about this if you want, but that's the TL;DR :)